### PR TITLE
Stop control characters being printed to terminal

### PR DESCRIPTION
**Pull request by tpoliaw**
Detail of change is in the commit message but easiest steps to reproduce is to open a new terminal and run the following

```sh
echo one
echo '^[[31mtwo' # with `^[` entered as literal using ctrl-v + ctrl-[
echo three
atuin history list -s
```

With the previous behaviour, the control sequence would not be visible (and therefore not copyable) and everything after the second 'echo' would be red.

Running `atuin search -i --filter-mode session` afterwards would display the `echo two` command as `echo '[31mtwo'`. Selecting it would print `echo 'two'` without the control sequence and two onwards in red.

### CODE REVIEW

File: atuin-common/src/utils.rs
Comment (akinomyoga) [inline atuin-common/src/utils.rs:122]: This might be just a preference, but is there a possibility to use the caret notation instead of the hex notation `\xHH`? I'm more familiar with the caret notation in the command line and thus the command strings (as the shells' line editors show the control characters in the caret notation). (Note: I'm not a maintainer, so I don't have any power. If you don't like this, you can ignore it.)

```suggestion
                buf.push('^');
                buf.push(match part.as_bytes()[i] {
                    0x7F => '?',
                    code => char::from_u32(u32::from(code) + 64).unwrap(),
                });
```

**Review by akinomyoga — APPROVED** [review]
This fix is also good for the multiline commands. Previously, lines were connected without any separators in the history list, so the structural information of the multiline commands was lost (or at least, was harder to grasp at a glance). Now the lines are separated by the escaped newlines, and these are visible. This is nice.

File: atuin-common/src/utils.rs
Comment (ellie) [reply atuin-common/src/utils.rs:122]: +1 here

**Comment by ellie** [comment]
Generally looks good to me! Thank you 🙏 

If we could get a `cargo fmt` for CI please

**Comment by tpoliaw** [comment]
(it looked like other PRs had used the force push approach so I copied them)

File: atuin-common/src/utils.rs
Comment (akinomyoga) [reply atuin-common/src/utils.rs]: > but the main problem with using `^[` etc is that you can't copy/paste the output to re-run the command as they'll be interpreted as two characters instead of one.

Ah, OK. I now understand the third paragraph of the cover of this PR. Maybe, is `echo '[31mtwo'` in the cover a typo of `echo '\x1b[31mtwo'`?

But then, I'm afraid to say that the hex notation still has the same problem for copying and pasting as the caret notation. The raised example of `echo` works with the hex notation because of the specific behavior of a command of a specific shell (i.e., echo in Bash and Fish doesn't work that way). The hex notation is not handled at the shell level, so we cannot expect it to work in general. I think there are even opposite cases, where the caret notation works while the hex notation fails. For example, in Zsh, `bindkey ^R some_widget` works, but `bind \x0D some_widget` doesn't work as expected.

In the first place, the users can also copy and paste the commands in the terminal's scrollback buffer, where the control characters in the commands are presented with the caret notation. I'm not sure if using the hex notation in Atuin would be actually useful in solving the essential issue. If a user wants to include a raw control character in commands, I feel the user finally needs to learn that copying and pasting the displayed command doesn't work in general.

**Comment by ellie** [comment]
> Ellie, how do you prefer to handle updating PRs, push new noisy commits and squash at the end (assuming it's accepted of course), or update the same commit and force push to the branch? (or something else?)

I'll update the contributing guide to clarify, but as we squash-and-merge I'm happy for people to do whatever they wish within the PR itself. It's much easier that way vs policing how people build up their work

**Review by ellie — APPROVED** [review]
Excellent! Thank you for working on this 🙏 

Seeing as this is your first time contributing, if you would like a holographic contributors-only Atuin sticker, then please fill out this [form!](https://notionforms.io/forms/contributors-stickers)

We do also have a [Discord](https://discord.gg/Fq8bJSKPHh) if you'd like to ask any questions, or just fancy hanging out!

File: atuin-common/src/utils.rs
Comment (tpoliaw) [reply atuin-common/src/utils.rs]: I think I'm coming round to this idea more now. It's less convenient for the cases I had that originally come across but I can see the `^[` version is more general and better matches the way the terminal displays control chars. As Ellie initially `+1`ed your original comment I'll change it now. Is it worth inverting the colours for escape characters (similar to how zsh does it) to distinguish it from the literal `^` + `[`?

Comment (akinomyoga) [reply atuin-common/src/utils.rs]: > Is it worth inverting the colours for escape characters (similar to how zsh does it) to distinguish it from the literal `^` + `[`?

Yeah, that sounds better. However, we might need adjustments for the ongoing PR #1574, which uses the terminal's reverse video for another meaning. In principle, we can treat *(reverse + reverse)* as *not-reverse*, but the terminal control sequences do not offer such a model. Someone needs to implement it in the application layer, but I'm not sure if that is already implemented in any of the Rust libraries that Atuin currently depends on.

Comment (tpoliaw) [reply atuin-common/src/utils.rs]: Ok, I'll leave the reverse for now, it can be figured out whenever 1574 is merged.

Comment (akinomyoga) [reply atuin-common/src/utils.rs]: > > It could be possible to use caret notation for the interactive search but the hex version when things are printed to stdout but I'm not sure the inconsistency would be worth it.
> 
> I think we usually switch the output format depending on whether `stdout` (or the destination stream) is connected to a TTY. We can escape the control characters when `stdout` is connected to TTY, or otherwise, output the raw command string.

Do you have a plan to fix this issue? The current `main` branch doesn't work for the multiline command:

image

`\x0a` is not treated as a newline of the shell syntax.

Comment (tpoliaw) [reply atuin-common/src/utils.rs]: Then yes, looks like I need to. Would it be better to add the tty check or should the search result always be printed as is? Or a flag could switch between the two (although I'd accept if there's a reluctance to add more flags to search)?

Comment (akinomyoga) [reply atuin-common/src/utils.rs]: Hmm, I leave it to you. It's similar to the coloring of the output of the commands. Some commands do not support colors at all. Some commands enable colors with the check for TTY through `--color=auto`. Some commands perform the TTY switching by default. All are understandable.

Comment (tpoliaw) [reply atuin-common/src/utils.rs]: #1585 and #1586 for these two changes.

Comment (ellie) [reply atuin-common/src/utils.rs]: are they both ok for review? I see they're marked as drafts

Comment (tpoliaw) [reply atuin-common/src/utils.rs]: Sure, I just figured I should test these changes more thoroughly than I did for this one and didn't have time when I submitted them. Given main is currently broken though I thought it was worth getting something in.

Comment (tpoliaw) [reply atuin-common/src/utils.rs]: Thanks, I'll have a look at what changed. And I will do but it might not be for a while.

**Pull request merged**
